### PR TITLE
Prevent ThreadContext header leak when sending response

### LINKED ISSUE #68278: org.elasticsearch.upgrades.RecoveryIT fails during upgrade tests
**Build scan**:
[6.8.14] https://gradle-enterprise.elastic.co/s/zrc4fnpfqnwdu

**Repro line**:
```
REPRODUCE WITH: ./gradlew ':qa:rolling-upgrade:v6.8.14#upgradedClusterTest' \
  -Dtests.class="org.elasticsearch.upgrades.RecoveryIT" \
  -Dtests.method="testRecoveryWithConcurrentIndexing" \
  -Dtests.seed=9A9A65701543F0D \
  -Dtests.security.manager=true \
  -Dtests.bwc=true \
  -Dtests.locale=es-VE \
  -Dtests.timezone=Arctic/Longyearbyen \
  -Druntime.java=8

REPRODUCE WITH: ./gradlew ':qa:rolling-upgrade:v6.8.14#upgradedClusterTest' \
  -Dtests.class="org.elasticsearch.upgrades.RecoveryIT" \
  -Dtests.method="testRecoveryWithConcurrentIndexing" \
  -Dtests.seed=9A9A65701543F0D \
  -Dtests.security.manager=true \
  -Dtests.bwc=true \
  -Dtests.locale=es-VE \
  -Dtests.timezone=Arctic/Longyearbyen \
  -Druntime.java=8
```

**Reproduces locally?**:
Don't know yet.

**Applicable branches**:
`7.x`, `7.10`

**Failure history**:

**Failure excerpt**:
```
12:46:20 org.elasticsearch.upgrades.RecoveryIT > testRecoveryWithConcurrentIndexing FAILED
12:46:20     org.elasticsearch.client.WarningFailureException: method [PUT], host [http://127.0.0.1:38975], URI [/recovery_with_concurrent_indexing/_settings], status line [HTTP/1.1 200 OK]
12:46:20     Warnings: ['Y' year-of-era should be replaced with 'y'. Use 'Y' for week-based-year.; 'Z' time zone offset/id fails when parsing 'Z' for Zulu timezone. Consider using 'X'. Use new java.time date format specifiers.]
12:46:20     {"acknowledged":true}
12:46:20         at __randomizedtesting.SeedInfo.seed([9A9A65701543F0D:89F8620483F0E1C7]:0)
12:46:20         at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:322)
12:46:20         at org.elasticsearch.client.RestClient.performRequest(RestClient.java:296)
12:46:20         at org.elasticsearch.client.RestClient.performRequest(RestClient.java:270)
12:46:20         at org.elasticsearch.test.rest.ESRestTestCase.updateIndexSettings(ESRestTestCase.java:1208)
12:46:20         at org.elasticsearch.test.rest.ESRestTestCase.updateIndexSettings(ESRestTestCase.java:1202)
12:46:20         at org.elasticsearch.upgrades.RecoveryIT.testRecoveryWithConcurrentIndexing(RecoveryIT.java:174)
```

### PR DESCRIPTION
**Pull request by pgomulka**
We need to stash thread context in `DefaultRestChannel` before we call `channel.sendResponse` because when calling this method, our execution might be delayed and the thread be reused for another task - like sending another response. And it would see thread context from the initial “delayed” work.

This commit also expands the assertions on the empty thread context to make sure it does not contains response headers.

closes https://github.com/elastic/elasticsearch/issues/68278

### CODE REVIEW

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Comment by pgomulka** [comment]
ok to test

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Comment by pgomulka** [comment]
TODO - add those assertions back https://github.com/elastic/elasticsearch/pull/84751/files#r821480480

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandler.java [hunk 1]
```diff
@@ -29,12 +32,17 @@
     private final Queue<WriteOperation> queuedWrites = new ArrayDeque<>();
 
     private WriteOperation currentWrite;
+    private ThreadPool threadPool;
 
-    public Netty4WriteThrottlingHandler() {}
+    public Netty4WriteThrottlingHandler(ThreadPool threadPool) {
```
Comment (DaveCTurner) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandler.java:37]: I'd rather capture just the `ThreadContext` itself here, not the whole `ThreadPool`.

**Review by DaveCTurner — COMMENTED** [review]
Looks good assuming CI is happy. I left one small request.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**